File: examples/csp_server_client.c
#include <csp/csp_debug.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>

#include <csp/csp.h>
#include <csp/drivers/usart.h>
#include <csp/drivers/can_socketcan.h>
#include <csp/interfaces/csp_if_zmqhub.h>

#include "csp_posix_helper.h"

/* Server port, the port the server listens on for incoming connections from the client. */
#define MY_SERVER_PORT		10

/* Commandline options */
static uint8_t server_address = 255;

/* test mode, used for verifying that host & client can exchange packets over the loopback interface */
static bool test_mode = false;
static unsigned int server_received = 0;
static unsigned int run_duration_in_sec = 3;

/* Server task - handles requests from clients */
void * server(void * param) {

	(void)param;

	csp_print("Server task started\n");

	/* Create socket with no specific socket options, e.g. accepts CRC32, HMAC, etc. if enabled during compilation */
	csp_socket_t sock = {0};

	/* Bind socket to all ports, e.g. all incoming connections will be handled here */
	csp_bind(&sock, CSP_ANY);

	/* Create a backlog of 10 connections, i.e. up to 10 new connections can be queued */
	csp_listen(&sock, 10);

	/* Wait for connections and then process packets on the connection */
	while (1) {

		/* Wait for a new connection, 10000 mS timeout */
		csp_conn_t *conn;
		if ((conn = csp_accept(&sock, 10000)) == NULL) {
			/* timeout */
			continue;
		}

		/* Read packets on connection, timout is 100 mS */
		csp_packet_t *packet;
		while ((packet = csp_read(conn, 50)) != NULL) {
			switch (csp_conn_dport(conn)) {
			case MY_SERVER_PORT:
				/* Process packet here */
				csp_print("Packet received on MY_SERVER_PORT: %s\n", (char *) packet->data);
				csp_buffer_free(packet);
				++server_received;
				break;

			default:
				/* Call the default CSP service handler, handle pings, buffer use, etc. */
				csp_service_handler(packet);
				break;
			}
		}

		/* Close current connection */
		csp_close(conn);

	}

	return NULL;

}
/* End of server task */

/* Client task sending requests to server task */
void * client(void * param) {

	(void)param;

	csp_print("Client task started\n");

	unsigned int count = 'A';

	while (1) {

		usleep(test_mode ? 200000 : 1000000);

		/* Send ping to server, timeout 1000 mS, ping size 100 bytes */
		int result = csp_ping(server_address, 1000, 100, CSP_O_NONE);
		csp_print("Ping address: %u, result %d [mS]\n", server_address, result);
        (void) result;

		/* Send reboot request to server, the server has no actual implementation of csp_sys_reboot() and fails to reboot */
		csp_reboot(server_address);
		csp_print("reboot system request sent to address: %u\n", server_address);

		/* Send data packet (string) to server */

		/* 1. Connect to host on 'server_address', port MY_SERVER_PORT with regular UDP-like protocol and 1000 ms timeout */
		csp_conn_t * conn = csp_connect(CSP_PRIO_NORM, server_address, MY_SERVER_PORT, 1000, CSP_O_NONE);
		if (conn == NULL) {
			/* Connect failed */
			csp_print("Connection failed\n");
			return NULL;
		}

		/* 2. Get packet buffer for message/data */
		csp_packet_t * packet = csp_buffer_get_always();

		/* 3. Copy data to packet */
        memcpy(packet->data, "Hello world ", 12);
        memcpy(packet->data + 12, &count, 1);
        memset(packet->data + 13, 0, 1);
        count++;

		/* 4. Set packet length */
		packet->length = (strlen((char *) packet->data) + 1); /* include the 0 termination */

		/* 5. Send packet */
		csp_send(conn, packet);

		/* 6. Close connection */
		csp_close(conn);
	}

	return NULL;
}
/* End of client task */

static void print_usage(void)
{
	csp_print("Usage:\n"
			  " -v <version>     set protocol version\n"
			  " -t               enable test mode\n"
			  " -T <duration>    enable test mode with running time in seconds\n"
			  " -h               print help\n");
}

/* main - initialization of CSP and start of server/client tasks */
int main(int argc, char * argv[]) {

    uint8_t address = 0;
    int opt;
    while ((opt = getopt(argc, argv, "v:tT:h")) != -1) {
        switch (opt) {
            case 'a':
                address = atoi(optarg);
                break;
            case 'r':
                server_address = atoi(optarg);
                break;
			case 'v':
				csp_conf.version = atoi(optarg);
				break;
            case 't':
                test_mode = true;
                break;
            case 'T':
                test_mode = true;
                run_duration_in_sec = atoi(optarg);
                break;
            case 'h':
				print_usage();
				exit(0);
                break;
            default:
				print_usage();
                exit(1);
                break;
        }
    }

    csp_print("Initialising CSP");

    /* Init CSP */
    csp_init();

    /* Start router */
    router_start();

    /* Add interface(s) */
    csp_iface_t * default_iface = NULL;
    if (!default_iface) {
        /* no interfaces configured - run server and client in process, using loopback interface */
        server_address = address;
    }

    csp_print("Connection table\r\n");
    csp_conn_print_table();

    csp_print("Interfaces\r\n");
    csp_iflist_print();

    /* Start server thread */
    csp_pthread_create(server);

    /* Start client thread */
    csp_pthread_create(client);

    /* Wait for execution to end (ctrl+c) */
    while(1) {
        sleep(run_duration_in_sec);

        if (test_mode) {
            /* Test mode is intended for checking that host & client can exchange packets over loopback */
            if (server_received < 5) {
                csp_print("Server received %u packets\n", server_received);
                exit(1);
            }
            csp_print("Server received %u packets\n", server_received);
            exit(0);
        }
    }

    return 0;
}

File: examples/csp_server_client.py
"""
Usage: LD_LIBRARY_PATH=build PYTHONPATH=build python3 ./examples/csp_server_client.py
"""
import time
import threading
import libcsp_py3 as csp
from typing import Any, Callable


def printer(node: str, color: str) -> Callable:
    def f(inp: str) -> None:
        print('{color}[{name}]: {inp}\033[0m'.format(
            color=color, name=node.upper(), inp=inp))

    return f


def server_task(addr: int, port: int) -> None:
    _print = printer('server', '\033[96m')
    _print('Starting server task')

    sock = csp.socket()
    csp.bind(sock, csp.CSP_ANY)

    csp.listen(sock, 10)

    while 1:
        conn = csp.accept(sock, 10000)

        if conn is None:
            continue

        while (packet := csp.read(conn, 50)) is not None:
            if csp.conn_dport(conn) == port:
                _print('Recieved on {port}: {data}'.format(
                    port=port,
                    data=csp.packet_get_data(packet).decode('utf-8'))
                )
            else:
                csp.service_handler(conn, packet)


def client_task(addr: int, port: int) -> None:
    _print = printer('client', '\033[92m')
    _print('Starting client task')

    count = ord('A')

    while 1:
        time.sleep(1)

        ping = csp.ping(addr, 1000, 100, csp.CSP_O_NONE)
        _print('Ping {addr}: {ping}ms'.format(addr=addr, ping=ping))

        conn = csp.connect(csp.CSP_PRIO_NORM, addr, port, 1000, csp.CSP_O_NONE)
        if conn is None:
            raise Exception('Connection failed')

        packet = csp.buffer_get(0)
        if packet is None:
            raise Exception('Failed to get CSP buffer')

        data = bytes('Hello World {}'.format(chr(count)), 'ascii') + b'\x00'
        count += 1

        csp.packet_set_data(packet, data)
        csp.send(conn, packet)


def main() -> None:
    csp.init("", "", "")
    csp.route_start_task()

    serv_addr = 0
    serv_port = 10

    for task in (server_task, client_task):
        t = threading.Thread(target=task, args=(serv_addr, serv_port))
        t.start()

    print('Server and client started')


if __name__ == '__main__':
    main()
